File: src/main/java/info/skyblond/archivedag/service/impl/ConfigService.java
package info.skyblond.archivedag.service.impl;

import info.skyblond.archivedag.model.entity.ConfigEntity;
import info.skyblond.archivedag.repo.ConfigRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

@Slf4j
@Service
public class ConfigService {
    private final ConfigRepository configRepository;

    public ConfigService(ConfigRepository configRepository) {
        this.configRepository = configRepository;
    }

    public Map<String, String> listConfig(String keyPrefix, Pageable pageable) {
        Map<String, String> result = new HashMap<>();
        this.configRepository.findAllByKeyStartingWith(keyPrefix.toLowerCase(), pageable)
                .forEach(r -> result.put(r.getKey(), r.getValue()));
        return Collections.unmodifiableMap(result);
    }

    @Transactional
    public void updateConfig(String key, String value) {
        ConfigEntity entity = new ConfigEntity(key.toLowerCase(), value);
        this.configRepository.save(entity);
    }

    public static final String ALLOW_GRPC_WRITE_KEY = "archive-dag.grpc.allow-write";

    public boolean allowGrpcWrite() {
        ConfigEntity entity = this.configRepository.findByKey(ALLOW_GRPC_WRITE_KEY);
        if (entity == null) {
            return false;
        }
        return Boolean.parseBoolean(entity.getValue());
    }
}

File: backend/src/main/kotlin/info/skyblond/archivedag/arstue/ConfigService.kt
package info.skyblond.archivedag.arstue

import info.skyblond.archivedag.commons.service.EtcdSimpleConfigService
import io.etcd.jetcd.ByteSequence
import io.etcd.jetcd.options.GetOption
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import java.util.*

@Service
class ConfigService(
    private val etcdConfig: EtcdSimpleConfigService,
) {
    private val etcdConfigPrefix = "/application/arstue/config/archive_dag/"

    fun listConfig(keyPrefix: String, pageable: Pageable): Map<String, String?> {
        require(pageable.offset <= Int.MAX_VALUE) { "Offset overflow" }

        val result: MutableMap<String, String?> = HashMap()
        etcdConfig.getConfig(
            ByteSequence.from(etcdConfig.getStringKey(etcdConfigPrefix, keyPrefix).lowercase(), Charsets.UTF_8),
            GetOption.newBuilder().isPrefix(true).withKeysOnly(true).build()
        ).kvs
            .map { it.key.toString().removePrefix(etcdConfigPrefix) }
            .sorted()
            .drop(pageable.offset.toInt())
            .take(pageable.pageSize)
            .forEach {
                val str = etcdConfig.getConfig(etcdConfigPrefix, it)
                result[it] = if (str.isNullOrBlank()) null else str
            }
        return Collections.unmodifiableMap(result)
    }

    fun updateConfig(key: String, value: String?) {
        etcdConfig.putConfig(etcdConfigPrefix, key, value ?: "")
    }

    fun allowGrpcWrite(): Boolean {
        return etcdConfig.getConfig(etcdConfigPrefix, ALLOW_GRPC_WRITE_KEY)?.toBooleanStrictOrNull() ?: false
    }

    companion object {
        const val ALLOW_GRPC_WRITE_KEY = "grpc.allow-write"
    }
}
